src/els_diagnostics_provider.erl

```diff
@@ -0,0 +1,111 @@
+-module(els_diagnostics_provider).
+
+-behaviour(els_provider).
+
+-export([ handle_info/2
+        , handle_request/2
+        , init/0
+        , is_enabled/0
+        , options/0
+        ]).
+
+-export([ notify/2
+        , publish/2
+        ]).
+
+-include("erlang_ls.hrl").
+
+-type state() :: #{in_progress => [progress_entry()]}.
+-type progress_entry() :: #{ uri := uri()
+                           , pending := [job()]
+                           , diagnostics := [els_diagnostics:diagnostic()]
+                           }.
+-type job() :: pid().
+
+-define(SERVER, ?MODULE).
+
+%%==============================================================================
+%% els_provider functions
+%%==============================================================================
+
+-spec is_enabled() -> boolean().
+is_enabled() -> true.
+
+-spec options() -> map().
+options() ->
+  #{}.
+
+-spec init() -> state().
+init() ->
+  #{ in_progress => [] }.
+
+%% LSP 3.15 introduce versioning for diagnostics. Until all clients
+%% support it, we need to keep track of old diagnostics and re-publish
+%% them every time we get a new chunk.
+-spec handle_info(any(), state()) -> state().
+handle_info({diagnostics, Diagnostics, Job}, State) ->
+  lager:debug("Received diagnostics [job=~p]", [Job]),
+  #{ in_progress := InProgress } = State,
+  { #{ pending := Jobs
+     , diagnostics := OldDiagnostics
+     , uri := Uri
+     }
+  , Rest
+  } = find_entry(Job, InProgress),
+  NewDiagnostics = Diagnostics ++ OldDiagnostics,
+  ?MODULE:publish(Uri, NewDiagnostics),
+  case lists:delete(Job, Jobs) of
+    [] ->
+      State#{in_progress => Rest};
+    Remaining ->
+      State#{in_progress => [#{ pending => Remaining
+                              , diagnostics => NewDiagnostics
+                              , uri => Uri
+                              }|Rest]}
+  end;
+handle_info(_Request, State) ->
+  State.
+
+-spec handle_request(any(), state()) -> {any(), state()}.
+handle_request({run_diagnostics, Params}, State) ->
+  #{in_progress := InProgress} = State,
+  #{ <<"textDocument">> := #{ <<"uri">> := Uri}} = Params,
+  lager:debug("Starting diagnostics jobs [uri=~p]", [Uri]),
+  Jobs = els_diagnostics:run_diagnostics(Uri),
+  Entry = #{uri => Uri, pending => Jobs, diagnostics => []},
+  {noresponse, State#{in_progress => [Entry|InProgress]}}.
```

Comment:
If I'm understanding the code right, if there is an ongoing group of diagnostics running for `Uri1` and a new group of diagnostics is triggered for the same `Uri1` we could end up with a weird situation where, one group and the other gets published alternatively, right?

What I'm trying to ask is why not make the `in_progress` value a map keyed on `uri`? Checking first if there is an existing entry for `Uri` and if there is:
1.  Send an empty diagnostics list.
2. Kill all running diagnostics jobs for that `Uri`.
3. Start the new group of diagnostics jobs.

Reply:
True. I would prefer to tackle this as a follow-up, though. Created #619 for this purpose.
We also need to add support for "cancellation" for background jobs, which could be used for this.
Also, `3.15` of the protocol adds a `version` field which should help, too.